core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -216,6 +218,11 @@ class ReplicaStateMachine(config: KafkaConfig,
           logSuccessfulTransition(replicaId, partition, replicaState(replica), OfflineReplica)
           replicaState.put(replica, OfflineReplica)
         }
+
+        replicasWithoutLeadershipInfo.foreach { replica =>
+          logSuccessfulTransition(replicaId, replica.topicPartition, replicaState(replica), OfflineReplica)
+          replicaState.put(replica, OfflineReplica)
+        }
```

Comment:
It seems that we could just iterate on validReplicas and get rid of the code from line 217 to 219?

Reply:
If we do that, then the replicas with leadership, but failed to be removed from ISR would also transition to OfflineReplica state. The removeReplicasFromIsr already has logic to mark those replica transitions as failed, so doing that would generate confusing logs.